### ISSUE TITLE
Import Sound Library not working for Windows 64 Bit

### ISSUE THREAD

**Issue by mikotoiii** [hydrogen-music/hydrogen#284]
As the title says, importing a sound library does not work on the 64 bit version of Windows version 0.9.7 (Which I have now got cross compiling). I think that it might have something to do with an edit I had to make to the source in src/gui/src/SongEditor.cpp at line 1475
INFOLOG( QString("[patternPopup_delete] Delete pattern: %1 @%2").arg(pattern->get_name()).arg( (long long)pattern ) );

When I tried to compile the 64 bit version, the compiler threw an error basically saying that there was a type mismatch (long should be long long from my understanding). I'm hoping that someone who works more with the core system can have a look and see if there's something that we can do, or if that would even cause the issue I'm having.

For reference here's a link to the branch I'm working from:
https://github.com/mikotoiii/hydrogen/tree/win_64/

**Comment by mikotoiii** [comment]
Here's the log error I get when the INFOLOG is set to long on 64 Bit.

# Output:
[ 20%] Built target hydrogen-core-0.9.7
[ 20%] Built target h2cli
[ 20%] Built target h2player
[ 21%] Built target h2synth
Scanning dependencies of target hydrogen
[ 22%] Building CXX object src/gui/CMakeFiles/hydrogen.dir/src/SongEditor/SongEditor.cpp.obj
In file included from /home/sboyer/build/hydrogen/src/core/include/hydrogen/basics/song.h:32:0,
                 from /home/sboyer/build/hydrogen/src/gui/src/SongEditor/SongEditor.cpp:27:
/home/sboyer/build/hydrogen/src/gui/src/SongEditor/SongEditor.cpp: In member function 'void SongEditorPatternList::deletePatternFromList(QString, QString, int)':
/home/sboyer/build/hydrogen/src/gui/src/SongEditor/SongEditor.cpp:1475:103: error: cast from 'H2Core::Pattern*' to 'long int' loses precision [-fpermissive]
  INFOLOG( QString("[patternPopup_delete] Delete pattern: %1 @%2").arg(pattern->get_name()).arg( (long)pattern ) );
                                                                                                       ^
/home/sboyer/build/hydrogen/src/core/include/hydrogen/object.h:109:141: note: in definition of macro '__LOG_METHOD'
 #define __LOG_METHOD(   lvl, msg )  if( __logger->should_log( (lvl) ) )                 { __logger->log( (lvl), class_name(), __FUNCTION__, msg ); }
                                                                                                                                             ^
/home/sboyer/build/hydrogen/src/gui/src/SongEditor/SongEditor.cpp:1475:2: note: in expansion of macro 'INFOLOG'
  INFOLOG( QString("[patternPopup_delete] Delete pattern: %1 @%2").arg(pattern->get_name()).arg( (long)pattern ) );
  ^
src/gui/CMakeFiles/hydrogen.dir/build.make:1120: recipe for target 'src/gui/CMakeFiles/hydrogen.dir/src/SongEditor/SongEditor.cpp.obj' failed
make[2]: *** [src/gui/CMakeFiles/hydrogen.dir/src/SongEditor/SongEditor.cpp.obj] Error 1
CMakeFiles/Makefile2:337: recipe for target 'src/gui/CMakeFiles/hydrogen.dir/all' failed
make[1]: *** [src/gui/CMakeFiles/hydrogen.dir/all] Error 2
Makefile:137: recipe for target 'all' failed
make: *** [all] Error 2

**Comment by mauser** [comment]
No, the INFOLOG function is just there for logging, it has no effect on the import of a drumkit. But importing a drumkit means that a file has to decompressed, and this is sth. which may go wrong.. Could you paste the output of hydrogen here when you're trying to import a drumkit?

**Comment by mikotoiii** [comment]
Here's the long from Hydrogen when trying to import:

```
(E) Filesystem::path_usable unable to create user directory : C:/Program Files/H
ydrogen/hydrogen/data←[0m
(E) XMLDoc::read XML document C:/Program Files/Hydrogen/data/drumkits/GMkit/drum
kit.xml is not valid (C:/Program Files/Hydrogen/data/xsd/drumkit.xsd), loading m
ay fail←[0m
(E) Legacy::load_drumkit loading drumkit with legacy code←[0m
(E) XMLDoc::read XML document C:/Program Files/Hydrogen/data/drumkits/TR808Emula
tionKit/drumkit.xml is not valid (C:/Program Files/Hydrogen/data/xsd/drumkit.xsd
), loading may fail←[0m
(E) Legacy::load_drumkit loading drumkit with legacy code←[0m
(E) HydrogenApp::showInfoSplash [showInfoSplash] Directory C:/Program Files/Hydr
ogen/data/doc/infoSplash not found.←[0m
(E) Filesystem::check_permissions C:/Program Files/Hydrogen/hydrogen/data/cache/
repositories/e7101e0b1e49a9e52574a2557163215b is not a file←[0m
(E) SoundLibraryImportDialog::writeCachedData Failed to open file for writing re
pository cache.←[0m
(E) XMLDoc::read XML document C:/Program Files/Hydrogen/data/drumkits/GMkit/drum
kit.xml is not valid (C:/Program Files/Hydrogen/data/xsd/drumkit.xsd), loading m
ay fail←[0m
(E) Legacy::load_drumkit loading drumkit with legacy code←[0m
(E) XMLDoc::read XML document C:/Program Files/Hydrogen/data/drumkits/TR808Emula
tionKit/drumkit.xml is not valid (C:/Program Files/Hydrogen/data/xsd/drumkit.xsd
), loading may fail←[0m
(E) Legacy::load_drumkit loading drumkit with legacy code←[0m
```

Also, as an aside, it doesn't seem to pull the online list of kits or patterns.

**Comment by mikotoiii** [comment]
Hmm, this seems to be related to permissions. I ran Hydrogen as Administrator on Windows, and it let me install the kit fine, and displayed the list of kits from the web. It even keeps the kit installed for users after the Administrator version closes and is run as a regular user again.

**Comment by mauser** [comment]
Correct, this seems to be the problem: (

E) Filesystem::path_usable unable to create user directory : C:/Program Files/Hydrogen/hydrogen/data

But i'm wondering where the difference between 32bit and 64bit is? Have you tried the same steps with the 32bit version?

**Comment by mikotoiii** [comment]
Hydrogen doesn't exhibit this behaviour in my 32 bit Windows 7 VM. Just on my Windows 7 64 bit host OS. I'll try uninstalling the 64 bit version on my host OS and install the 32 bit version to see if the problem persists.

**Comment by mikotoiii** [comment]
No go, the 32 bit version on the 64 bit OS works as expected. This seems limited to the 64 bit version specifically.

**Comment by mikotoiii** [comment]
So after digging through a lot of stuff, I think I have figured out what is happening. As for a solution, I haven't come up with anything that has worked yet, but I'm still working on it.

The trouble boils down to Windows UAC blocking the program from editing the folder in the Program Files directory, which is where Hydrogen is trying to download the drumkits to. This is bypassed if you specifically run the program as Administrator, or give RW permissions to the Users group on the folder itself in Program Files. The problem with this is that neither of these "solutions" are standard practice, or even all that great.

What I've been looking at is embedding UAC prompting / tokens into the rc file I created for the icon, which should in theory let the user/program have RW to the folder, and not the virtual space that normal / older applications live. In practice however, this has not been working for me (the rc file is embedding the manifest properly into the exe, but even with requireAdministrator set, it still won't prompt for UAC elevation). This one may take some time to figure out.
